Articles/Blog/WhenIsBetterToWriteAutomatedTests-src.md

```diff
@@ -0,0 +1,110 @@
+
+- <img src='https://github.com/betterscientificsoftware/images/raw/master/testing-hero-image.png'/>[Image Source: https://www.testingspot.net (see below)]
+
+#### Contributed by [Roscoe A. Barltett](https://bartlettroscoe.github.io)
+#### Publication date: MONTH DD, YEAR
+
+*SOME GOOD SUBTITLE.  Any suggestions?*
+
+<br>
+
+## Introduction
+
+The importance of writing and maintaining strong automated tests for software is well established in the modern software engineering community (and in fact, tests will be written before the code when using test-driven development, TDD).
+But does it always pay off to write automated tests for some piece of code?
+Are there situations where one is better off not writing automated tests?
+Will you be considered to be a bad developer and chastised by your peers, users, stakeholders, or your manager if every line of code that you write is not under strong automated testing?
+
+I have been wrestling with this question since 2007 when I first read the book "Working Effectively with Legacy Code"<sup>[1]</sup> and radically changed my thinking about automated testing.
+In the 10+ years of my professional career prior to 2007, I wrote a lot of Computational Science and Engineering (CSE) and applied math software with very little strong automated testing.
+I used the validation-centric approach<sup>[2]</sup> to test the software that I was writing (i.e. only test the software in the downstream customer code and then parhaps only manually).
+As a result, I had personally written O(100k) lines of code that were being used by other people and was in the position to have to maintain.
+When I realized that almost all of this software was considered "Legacy Code" according to "Working Effectively with Legacy Code" (i.e. "code without tests == legacy code"), I felt ashamed and horrified at the mass of useful but buggy and difficult-to=maintain software that I had created and now would have to deal with for years to come.
+(Just recently a nasty bug came up in software I wrote around 2006 without strong automated testing that caused a good deal of harm and was an embarrassment.)
+
+After reading the book "Working Effectively with Legacy Code", I resolved to turn over a new leaf and become a born-again Agile software engineer that would use TDD to write all new code and religiously apply the Legacy Code Change Algorithm to modify all legacy code (i.e. code without tests).
+Over the years since my conversion to an Agile software engineer, I often rigorously wrote automated tests for nearly every piece of software I touched.
+There where times  under schedule pressure that I slipped and reverted back to my old ways and failed to write sufficient automated tests and came to regret it.
+However, there were also times where my overzealous drive to religiously test everything (which usually came after the guilt of a recent slip of not testing enough) where I wasted time writing automated tests that were not worth the effort.
```

Comment:
> where I wasted time writing automated tests that were not worth the effort.
Personal examples, if easy to explain, will strengthen the article

Reply:
> > where I wasted time writing automated tests that were not worth the effort.
>
> Personal examples, if easy to explain, will strengthen the article

If I am being honest, there are not many examples where I regretted having written automated tests.  One example is that I think I spent too much time writing code for the Trilinos package GlobiPack that did not end up getting a lot of usage and has now been deprecated and to be removed from Trilinos.  I think it is more honest to say that clearly defining these 5 criteria has helped to remove some guilt and general ill that I felt for not writing automated tests for some scripts that I have written in the last few years.  The ill feeling in my stomach has gone away from some of these (because they match the 5 criteria) but now I realize that I need to write some automated tests from some other bits of bash code that does not match all 5 of these criteria.  This has really helped to focus my decision making process for when to not write automated tests.  Hopefully this will help other people as well.

Reply:
Tried to give an example in https://github.com/betterscientificsoftware/betterscientificsoftware.github.io/pull/377/files#diff-ff08f44603f4e10f3095baf6cea0da6eR31.